destination/destination.go

```diff
@@ -0,0 +1,360 @@
+func (d *Destination) writeInsertOperation(data *bytes.Buffer, item sdk.Record) error {
+	jsonEncoder := json.NewEncoder(data)
+
+	// Prepare data
+	metadata, payload, err := d.client.PrepareCreateOperation(item)
+	if err != nil {
+		return fmt.Errorf("failed to prepare metadata: %w", err)
+	}
+
+	// Write metadata
+	if err := jsonEncoder.Encode(metadata); err != nil {
+		return fmt.Errorf("failed to prepare metadata: %w", err)
+	}
+
+	// Write payload
+	if err := jsonEncoder.Encode(payload); err != nil {
+		return fmt.Errorf("failed to prepare data: %w", err)
+	}
+
+	return nil
+}
+
+// writeUpsertOperation adds upsert a Document with ID request into Bulk API request
+func (d *Destination) writeUpsertOperation(key string, data *bytes.Buffer, item sdk.Record) error {
+	jsonEncoder := json.NewEncoder(data)
+
+	// Prepare data
+	metadata, payload, err := d.client.PrepareUpsertOperation(key, item)
+	if err != nil {
+		return fmt.Errorf("failed to prepare metadata with key=%s: %w", key, err)
+	}
+
+	// Write metadata
+	if err := jsonEncoder.Encode(metadata); err != nil {
+		return fmt.Errorf("failed to prepare metadata with key=%s: %w", key, err)
+	}
+
+	// Write payload
+	if err := jsonEncoder.Encode(payload); err != nil {
+		return fmt.Errorf("failed to prepare data with key=%s: %w", key, err)
+	}
+
+	return nil
+}
+
+// writeDeleteOperation adds delete a Document by ID request into Bulk API request
+func (d *Destination) writeDeleteOperation(key string, data *bytes.Buffer) error {
+	jsonEncoder := json.NewEncoder(data)
+
+	// Prepare data
+	metadata, err := d.client.PrepareDeleteOperation(key)
+	if err != nil {
+		return fmt.Errorf("failed to prepare metadata with key=%s: %w", key, err)
+	}
+
+	// Write metadata
+	if err := jsonEncoder.Encode(metadata); err != nil {
+		return fmt.Errorf("failed to prepare metadata with key=%s: %w", key, err)
+	}
+
+	return nil
+}
+
+// executeBulkRequest executes Bulk API request and parses the response
+func (d *Destination) executeBulkRequest(ctx context.Context, data *bytes.Buffer) (bulkResponse, error) {
+	// Check if there is any job to do
+	if data.Len() < 1 {
+		sdk.Logger(ctx).Info().Msg("no operations to execute in bulk, skipping")
+
+		return bulkResponse{}, nil
+	}
+
+	defer data.Reset()
+
+	// Execute the request
+	responseBody, err := d.client.Bulk(ctx, bytes.NewReader(data.Bytes()))
+	if err != nil {
+		return bulkResponse{}, fmt.Errorf("bulk request failure: %w", err)
+	}
+
+	// Get the response
+	bodyContents, err := io.ReadAll(responseBody)
+	if err != nil {
+		return bulkResponse{}, fmt.Errorf("bulk response failure: failed to read the result: %w", err)
+	}
+	defer responseBody.Close()
+
+	// Read individual errors
+	var response bulkResponse
+	if err := json.Unmarshal(bodyContents, &response); err != nil {
+		return bulkResponse{}, fmt.Errorf("bulk response failure: could not read the response: %w", err)
+	}
```

Comment:
That's right 🙂 This needs to be changed because each client should be responsible for unmarshalling it's responses since they may differ from version to version. I got this on my to-do list.

Reply:
Noted https://github.com/miquido/conduit-connector-elasticsearch/issues/3